sklearn/ensemble/_stacking.py

```diff
@@ -152,14 +152,21 @@ def fit(self, X, y, sample_weight=None):
 
         stack_method = [self.stack_method] * len(all_estimators)
 
-        # Fit the base estimators on the whole training data. Those
-        # base estimators will be used in transform, predict, and
-        # predict_proba. They are exposed publicly.
-        self.estimators_ = Parallel(n_jobs=self.n_jobs)(
-            delayed(_fit_single_estimator)(clone(est), X, y, sample_weight)
-            for est in all_estimators
-            if est != "drop"
-        )
+        if self.cv == "prefit":
+            self.estimators_ = []
+            for estimator in all_estimators:
+                if estimator != "drop":
+                    check_is_fitted(estimator)
+                    self.estimators_.append(estimator)
```

Comment:
I am unsure about our API contract here. If someone alters an estimator from `all_estimators` without calling `fit` of the stacking model, it will however have an effect on the prediction.

Thinking about the freezing API, changing a hyperparameter and calling `fit` will have no effect. So the deep copy is not necessary to prevent the behaviour that I described above. However, without this freezing API here, we would need to make a deep copy of each estimator to prevent any side effects.

@thomasjpfan WDYT? I see that we have a similar pattern in calibration.

Reply:
Here is an example to explicitly illustrate what I mean:

```python
from sklearn.datasets import load_iris
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import LinearSVC
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.ensemble import StackingClassifier
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, stratify=y, random_state=42
)

X_train_1, X_train_2, y_train_1, y_train_2 = train_test_split(
    X_train, y_train, stratify=y_train, random_state=0, test_size=0.5,
)

X, y = load_iris(return_X_y=True)
estimators = [
    ('rf', RandomForestClassifier(n_estimators=10, random_state=42).fit(X_train_1, y_train_1)),
    ('svr', make_pipeline(StandardScaler(),
                          LinearSVC(random_state=42)).fit(X_train_1, y_train_1))
]

clf = StackingClassifier(
    estimators=estimators, final_estimator=LogisticRegression(), cv="prefit"
)

clf.fit(X_train, y_train)
print(f"Accuracy score: {clf.score(X_test, y_test):.2f}")

estimators[0][1].fit(X_train_2, y_train_2)

print(
    "Accuracy score after refitting an estimator: "
    f"{clf.score(X_test, y_test):.2f}"
)
```

```
Accuracy score: 0.89
Accuracy score after refitting an estimator: 0.95
```

Reply:
Basically, we already have an issue with `CalibratedClassifierCV`. So I assume that we will be in trouble in the future but we might want to solve those as a whole.